### ISSUE TITLE
Buiding caffe error: undefined reference to `cv::imread(cv::String const&, int)' 

### ISSUE THREAD

**Issue by sunnyxiaohu** [yjxiong/temporal-segment-networks#68]
Hi, when I run build_all.sh, I got the error below: 

..........................................

[ 98%] Scanning dependencies of target convert_mnist_siamese_data
Scanning dependencies of target pycaffe
Building CXX object examples/CMakeFiles/convert_mnist_data.dir/mnist/convert_mnist_data.cpp.o
[100%] Building CXX object examples/CMakeFiles/convert_mnist_siamese_data.dir/siamese/convert_mnist_siamese_data.cpp.o
[100%] Building CXX object python/CMakeFiles/pycaffe.dir/caffe/_caffe.cpp.o
Linking CXX executable mnist/convert_mnist_data
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/mnist/convert_mnist_data] Error 1
make[1]: *** [examples/CMakeFiles/convert_mnist_data.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Linking CXX executable compute_image_mean
Linking CXX executable cifar10/convert_cifar_data
Linking CXX executable device_query
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/compute_image_mean] Error 1
make[1]: *** [tools/CMakeFiles/compute_image_mean.dir/all] Error 2
[100%] Built target device_query
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/cifar10/convert_cifar_data] Error 1
make[1]: *** [examples/CMakeFiles/convert_cifar_data.dir/all] Error 2
Linking CXX executable siamese/convert_mnist_siamese_data
Linking CXX executable train_net
Linking CXX executable convert_imageset
Linking CXX executable upgrade_net_proto_binary
[100%] Linking CXX executable net_speed_benchmark
Built target train_net
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/siamese/convert_mnist_siamese_data] Error 1
make[1]: *** [examples/CMakeFiles/convert_mnist_siamese_data.dir/all] Error 2
[100%] Built target net_speed_benchmark
Linking CXX executable test_net
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/convert_imageset] Error 1
make[1]: *** [tools/CMakeFiles/convert_imageset.dir/all] Error 2
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/upgrade_net_proto_binary] Error 1
make[1]: *** [tools/CMakeFiles/upgrade_net_proto_binary.dir/all] Error 2
Linking CXX executable upgrade_net_proto_text
[100%] Built target test_net
Linking CXX executable finetune_net
[100%] Built target finetune_net
../lib/libcaffe.so: undefined reference to `cv::imread(cv::String const&, int)'
../lib/libcaffe.so: undefined reference to `cv::String::allocate(unsigned long)'
../lib/libcaffe.so: undefined reference to `cv::String::deallocate()'
../lib/libcaffe.so: undefined reference to `cv::imencode(cv::String const&, cv::_InputArray const&, std::vector<unsigned char, std::allocator<unsigned char> >&, std::vector<int, std::allocator<int> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/upgrade_net_proto_text] Error 1
make[1]: *** [tools/CMakeFiles/upgrade_net_proto_text.dir/all] Error 2

................................................................

Enviroment: 
1. python2.7.6(system itself) and annconda2.7.12（install path: ~/annoconda2）
2. opencv3.0.0(install before) and opencv(2.4.12)(install by running build_all.sh default)

Tried:
its seems that it can not find the corresponding libopencv_xx, so i just add the path by the following operation:
1. add "opencv-2.4.12/build/lib" to ~/bashrc
2. add "opencv-2.4.12/build/lib" to /etc/ld.so.conf.d/opencv.conf

However, it make no sense.

**Comment by yjxiong** [comment]
It is possible you compile with one version of OpenCV but link with the other one. Please refer to `build_all.sh` for build commands.

**Comment by sunnyxiaohu** [comment]
So, How can I known that which OpenCV version does it used when compiling, and when linking ?  For example, if the executable file is produced ( e.g. caffe ), then we could use command: "ldd caffe | grep opencv" to see which version did it used. 

Referd to you recommandation,  I tried the following steps:
1.  change cmake configurations commands as: 
OpenCV_DIR=../../../3rd-party/opencv-$version/build/ 
OpenCV_LIB_DIR=../../../3rd-party/opencv-$version/build/lib
cmake .. -DCUDA_USE_STATIC_CUDA_RUNTIME=OFF

2.  adjust the CMakeCache.txt built on caffe-action:
adding ------- OpenCV_LIB_DIR:PATH="$TSN_Root/3rd-party/opencv-$version/build/lib"

However, it make no sense. And I found that if the OpencvConfig.cmake could be found correctly, the corresponding LIB_PATH of Opencv also should be parsed correctly ?

Then, I compile dense_flow use OpenCV-2.4.12, and compile caffe-action use OpenCV-3.0.0, 
the project was build successfully.
I'm wondering if something else problem would be occurred ?
